### PR TITLE
test: add type hints to test_model

### LINKED ISSUE #1007: Run Pyright on test files to catch typing issues
We currently don't run Pyright over our `test/*.py` files to type-check those, meaning we're not getting "coverage" of most of our type annotations, which results in issues like https://github.com/canonical/operator/issues/963.

We should enable Pyright on these files. I suspect this will be a fairly big diff -- if so, we may want to opt the files in one by one in separate PRs to streamline review.

### PR DESCRIPTION
**Pull request by tonyandrewmeyer**
* One change to ops/charm, but only in the type hints. It looks from the use of `Required` that the other fields are not meant to be required, so make that the case (this also appears to match the code).
* Ignore types in the tests that are checking that bad types are handled properly.
* Help type checkers understand when some optional attributes won't be None.
* Replace a few bits of code where a minimal object was passed, enough for the test to do it's check, but not of the correct type, with an actual object of the proper type.
* Sprinkle type hints where required for pyright to understand what's happening.

Partially fixes #1007

### CODE REVIEW

File: test/test_model.py
Comment (benhoyt) [inline test/test_model.py:2082]: Similar question about naming: what do you think about `ensure_binding`?

File: test/test_model.py
Comment (benhoyt) [inline test/test_model.py:2558]: Not your issue, but IMO this block would be a whole lot simpler (and even shorter!) just written out the dumb way:

```python
fake_script(self, 'storage-list', 'echo fooerror >&2 ; exit 1')
with self.assertRaises(ModelError):
    self.backend.storage_list('foobar')
self.assertEqual(fake_script_calls(self, clear=True),
                 [['storage-list', 'foobar', '--format=json']])
```

Maybe you could make that change and it would avoid the messy type annotation too?

File: test/test_model.py
Comment (benhoyt) [inline test/test_model.py:2766]: Similar here: clearer and simpler to just write out the test cases. Again, not your issue, but might as well clean it up and avoid the messy type annotation.

I don't mind case-based tests (what Go calls "table-driven tests"), but I think they should only be used when there are really a lot of cases and the case-based approach is clearer.

File: test/test_model.py
Comment (benhoyt) [inline test/test_model.py:2785]: I think this is a case where case-based tests *do* improve things: the cases are short and there are a lot of them.

But maybe we can cheat with types here and just use `typing.Any` -- I think the type signature is just getting in our way here.

File: test/test_model.py
Comment (benhoyt) [inline test/test_model.py:2806]: Ditto here: `typing.Any`.

File: test/test_model.py
Comment (benhoyt) [inline test/test_model.py:3007]: Ditto here: `typing.Any` (and revert formatting to one case per line).

File: test/test_model.py
Comment (benhoyt) [inline test/test_model.py:3017]: Ditto here: `typing.Any`.

File: test/test_model.py
Comment (benhoyt) [inline test/test_model.py:3181]: Just wondering: do we have our tooling set up wrongly that it's doing awkward things like this? Is the formatter line break set to less than the linter line break or something?

```suggestion
                    content: typing.Optional[typing.Dict[str, str]] = None):
```

File: test/test_model.py
Comment (benhoyt) [inline test/test_model.py:3363]: Seeing we're probing into internals anyway, I wonder if we should just reuse the existing backend and cache objects?

```suggestion
        secret = self.make_secret(id='x')
        unit = ops.Unit('test', ops.CharmMeta(), self.model._backend, self.model._cache)
        rel123 = ops.Relation('test', 123, True, unit, self.model._backend, self.model._cache)
        rel234 = ops.Relation('test', 234, True, unit, self.model._backend, self.model._cache)
        secret.grant(rel123)
        unit = ops.Unit('app/0', ops.CharmMeta(), self.model._backend, self.model._cache)
        secret.grant(rel234, unit=unit)
```

**Review by benhoyt — COMMENTED** [review]
Nice work! Left a bunch of suggestions to simplify / clean a few things up.

File: test/test_model.py
Comment (tonyandrewmeyer) [reply test/test_model.py:69]: Sure. I was originally aiming for the change to be minimal (`model.get_relation` becoming `model_get_relation`) but a clearer name is better, I agree.

**Review by benhoyt — APPROVED** [review]
Thanks for the updates! I won't merge this until you make a final decision on those last outstanding comments -- your call as to whether to use `typing.Any` or whatever there, or keep the annotations.

File: test/test_model.py
Comment (tonyandrewmeyer) [reply test/test_model.py]: Agreed with the two for secrets; too much noise. I do somewhat like the ones for the list of metric, label pairs. I've cleaned it up a bit by defining an alias and just using that consistently. Work for you?

**Pull request merged**
